workers/gossip-server/src/peer_discovery/peers.rs

```diff
@@ -51,6 +52,50 @@ impl GossipProtocolExecutor {
         Ok(GossipResponse::Heartbeat(resp))
     }
 
+    /// Handle a proposed expiry from a peer
+    ///
+    /// TODO: Update this logic, for now we simply check if the local peer
+    /// thinks the expiry candidate should expire
+    pub async fn handle_propose_expiry(
+        &self,
+        sender: WrappedPeerId,
+        peer_id: WrappedPeerId,
+    ) -> Result<(), GossipError> {
+        let peer_info = self.global_state.get_peer_info(&peer_id).await?;
+        let info = match peer_info {
+            Some(info) => info,
+            None => return Ok(()),
+        };
+
+        // If the local peer has received a recent heartbeat from the candidate, notify
+        // the sender that the expiry should not proceed
+        let now = get_current_time_millis();
+        let time_since_last_heartbeat = now - info.last_heartbeat;
+        if time_since_last_heartbeat < HEARTBEAT_FAILURE_MS {
+            info!("rejecting expiry of {peer_id} from {sender}, last heartbeat was {time_since_last_heartbeat}ms ago");
+
+            // The peer should not expire yet, send a rejection
+            let msg = GossipRequest::RejectExpiry(RejectExpiryRequest {
+                peer_id,
+                last_heartbeat: info.last_heartbeat,
+            });
+
+            let job = NetworkManagerJob::request(sender, msg);
+            self.network_channel.send(job).map_err(err_str!(GossipError::SendMessage))?;
+        }
```

Comment:
If the peer does _not_ reject here, should they also follow through with expiry logic locally? Otherwise they have to wait for their own heartbeats with the expiry candidate to time out, though it seems semantically the decision to expire is made here